rts/src/gc.c

```diff
@@ -0,0 +1,234 @@
+  if (ptr == NULL) { return; }
+  Mark* location = (Mark*) ptr;
+  location--;
+  *location = gc_mark;
+}
+
+int is_marked_location(void* ptr) {
+  Mark* location = (Mark*) ptr;
+  if (location == NULL) { return 1; }
+  location--;
+  return *location == gc_mark;
+}
+
+int is_regular_closure(closure_t* closure) {
+  return closure->code != *tim_value_code &&
+         closure->code != *tim_nil_code;
+}
+
+void mark_frame_in_closure(closure_t* closure) {
+  if (is_regular_closure(closure)) {
+      mark_frame(closure->frame);
+    } else {
+      mark_location(closure->frame);
+    }
+}
+
+void mark_frame(frame_t frame) {
+  if (is_marked_location(frame)) { return; }
+  mark_location(frame);
+  mark_location(frame->arguments);
+  for (long i = 0; i < frame->length; i++) {
+    mark_frame_in_closure(&frame->arguments[i]);
+  }
+}
+
+void mark_closure(closure_t* closure) {
+  if (is_marked_location(closure)) { return; }
+  mark_location(closure);
+  mark_frame_in_closure(closure);
+}
+
+void mark_tim_metadata(tim_metadata_t metadata) {
+  if (is_marked_location(metadata)) { return; }
+  mark_location(metadata);
+  mark_frame(metadata->frame);
+}
+
+void mark_closure_stack(long size,stack_t stack) {
+  // Stack size correctness
+  assert( 0 < size || stack == NULL);
+  if (size <= 0) { return; }
+  assert(stack);
+  mark_closure(stack->data);
+  mark_closure_stack(size-1,stack->next);
+}
+
+void mark_closure_dump(dump_t dump) {
+  if (dump == NULL) { return; }
+  assert(dump);
+  stack_t current_stack = dump->current;
+  long stack_size = dump->current_size;
+  mark_closure_stack(stack_size,current_stack);
+  mark_tim_metadata(dump->metadata);
+  return mark_closure_dump(dump->parent);
+}
+
+void mark_argument_stack() {
+  mark_closure_dump(argument_stack);
+}
+
+void mark_value_stack (long size, stack_t stack) {
+  // Stack size correctness
+  assert( 0 < size || stack == NULL);
+  if (size <= 0) { return; }
+  assert(stack);
+  mark_location(stack->data);
+  mark_value_stack(size-1,stack->next);
+}
+
+void mark_value_dump(dump_t dump) {
+  if (dump == NULL) { return; }
+  assert(dump);
+  stack_t current_stack = dump->current;
+  long stack_size = dump->current_size;
+  mark_value_stack(stack_size,current_stack);
+  return mark_value_dump(dump->parent);
+}
+
+void mark_current_value_stack() {
+  mark_value_dump(value_stack);
+}
+
+void mark() {
+  mark_frame(current_frame);
+  mark_frame(current_data_frame);
+  mark_current_value_stack();
+  mark_argument_stack();
+}
+
+/************/
+/* Sweeping */
+/************/
+
+int is_marked_list_location(gc_list locations) {
+  return is_marked_location(locations->location);
+}
+
+void free_list_location(gc_list locations) {
+  free_location(locations->location);
+}
+
+void sweep_locations(gc_list parent, gc_list location) {
+  if (location == NULL) {return;}
+  assert(location);
+
+  if (is_marked_list_location(location)) {
+    return sweep_locations(location,location->next);
+  }
```

Comment:
The C standard specifies the evaluation order of `||` to always short circuit left-to-right, so we could simplify the checks above as:

```
if (location == NULL || is_marked_list_location(location)) { return; }
```

At the cost of making the code a bit more obscure.

Reply:
As the author of such hits as:
```
void specific_foo(){
  return foo()
}
```
I say we we keep it.